src/hotspot/share/gc/shared/workgroup.cpp

```diff
@@ -366,28 +366,32 @@ void SubTasksDone::clear() {
     _tasks[i] = 0;
   }
   _threads_completed = 0;
-#ifdef ASSERT
-  _claimed = 0;
-#endif
 }
 
-bool SubTasksDone::try_claim_task(uint t) {
-  assert(t < _n_tasks, "bad task id.");
-  uint old = _tasks[t];
-  if (old == 0) {
-    old = Atomic::cmpxchg(&_tasks[t], 0u, 1u);
-  }
-  bool res = old == 0;
+void SubTasksDone::all_tasks_completed_impl(uint n_threads,
+                                            uint skipped[],
+                                            size_t skipped_size) {
 #ifdef ASSERT
-  if (res) {
-    assert(_claimed < _n_tasks, "Too many tasks claimed; missing clear?");
-    Atomic::inc(&_claimed);
+  // all non-skipped tasks are claimed
+  for (uint i = 0; i < _n_tasks; ++i) {
+    if (_tasks[i] == 0) {
```

Comment:
This could be fixed in a separate CR: _tasks could be an array of bool instead of (u)int. Using an int is a historic artifact of not having a good Atomics library.

Reply:
Created a ticket (JDK-8259851) for it; will start working on that after this is merged. Thanks for the suggestion.